### PR TITLE
Job and better anonymization

### PR DESCRIPTION
**Pull request by gacou54**
- Add the `Job` class
- Allow more parameters in the `.anonymize()` methods
- Add an `.anonymize_as_job()` method that returns a job object instead of waiting for the anonymous patient.

related to #32

### CODE REVIEW

File: pyorthanc/_resources/patient.py [hunk 1]
```diff
@@ -247,24 +244,73 @@ def anonymize(self, remove: List = None, replace: Dict = None, keep: List = None
...
+            keep=['PatientName'],
+            replace={'PatientID': 'TheNewPatientID'},
+            force=True
+        )
+        ```
+        For large patient (recommended)
+        ```python
+        job = patient.anonymize(asynchronous=True)
+        job.state
+        job.block_until_completion()
```
Comment (gacou54) [inline pyorthanc/_resources/patient.py]: `block_until_completion()` -> `wait_until_completion()`

File: pyorthanc/jobs.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    running = 'Running'
+    failure = 'Failure'
+    pending = 'Pending'
+    paused = 'Paused'
+    retry = 'Retry'
+
+
+@dataclass
+class Error:
+    code: int
```
Comment (gacou54) [inline pyorthanc/jobs.py]: Replace the `Error` for the ErrorEnum from `orthanc-mock`

File: pyorthanc/_resources/patient.py [hunk 3]
```diff
@@ -224,16 +218,19 @@ def studies(self) -> List[Study]:
...
             return self._child_resources
 
-        studies_information = self.client.get_patients_id_studies(self.id_)
+        studies_ids = self.get_main_information()['Studies']
 
-        return [Study(i['ID'], self.client, self.lock) for i in studies_information]
+        return [Study(i, self.client, self.lock) for i in studies_ids]
 
-    def anonymize(self, remove: List = None, replace: Dict = None, keep: List = None, force: bool = False) -> 'Patient':
+    def anonymize(self, remove: List = None, replace: Dict = None, keep: List = None,
```
Comment (gacou54) [inline pyorthanc/_resources/patient.py:232]: Use an `anonymize` and an `anonymize_as_job`, when returning a `Patient` or a `Job` respectively.

File: pyorthanc/jobs.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    success = 'Success'
+    running = 'Running'
+    failure = 'Failure'
+    pending = 'Pending'
+    paused = 'Paused'
+    retry = 'Retry'
+
+
+@dataclass
+class Error:
```
Comment (gacou54) [inline pyorthanc/jobs.py]: Replace the `Error` for the ErrorEnum from `orthanc-mock`

File: pyorthanc/_resources/patient.py [hunk 5]
```diff
@@ -242,23 +253,136 @@ def anonymize(self, remove: List = None, replace: Dict = None, keep: List = None
...
+            'Priority': priority,
+            'Permissive': permissive,
+        }
+        if dicom_version is not None:
+            data['DicomVersion'] = dicom_version
+
+        try:
+            anonymous_patient = self.client.post_patients_id_anonymize(self.id_, data)
+        except ReadTimeout:
+            raise ReadTimeout('Patient is too long to process. Use `.anonymize_as_job` or increase client.timeout.')
```
Comment (ylemarechal) [inline pyorthanc/_resources/patient.py]: Maybe 'Patient anonymisation' ?

File: pyorthanc/_resources/patient.py [hunk 6]
```diff
@@ -242,23 +253,136 @@ def anonymize(self, remove: List = None, replace: Dict = None, keep: List = None
...
 
         return Patient(anonymous_patient['ID'], self.client)
 
+    def anonymize_as_job(self, remove: List = None, replace: Dict = None, keep: List = None,
+                         force: bool = False, keep_private_tags: bool = False,
+                         keep_source: bool = True, priority: int = 0, permissive: bool = False,
+                         dicom_version: str = None) -> Job:
+        """Anonymize patient and return a job
+
+        If no error has been raise, then it creates a new anonymous patient.
```
Comment (ylemarechal) [inline pyorthanc/_resources/patient.py]: raised ?

File: pyorthanc/_resources/patient.py (hunk 5 above)
Comment (gacou54) [reply pyorthanc/_resources/patient.py]: indeed

File: pyorthanc/_resources/patient.py (hunk 6 above)
Comment (gacou54) [reply pyorthanc/_resources/patient.py]: You're right! In the "job" version, we should not have an error if the job failed

File: pyorthanc/jobs.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+class State(Enum):
+    success = 'Success'
+    running = 'Running'
+    failure = 'Failure'
+    pending = 'Pending'
+    paused = 'Paused'
+    retry = 'Retry'
+
+
+class ErrorCode(Enum):
```
Comment (ylemarechal) [inline pyorthanc/jobs.py:19]: I would have ordered alphabetically

File: pyorthanc/jobs.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
+import time
+from datetime import datetime
+from enum import Enum
+from typing import Dict, Optional
+
+from . import util
+from .client import Orthanc
+
+
+class State(Enum):
```
Comment (ylemarechal) [inline pyorthanc/jobs.py:10]: I would order it alphabeticaly

**Review by ylemarechal — APPROVED** [review]

**Pull request merged**